### ISSUE TITLE
PhantomJS pageloads generate an extraneous POP event

### ISSUE THREAD

**Issue by shepmaster** [ReactTraining/history#511]
I'm noticing extra `POP` events generated when running tests inside of PhantomJS. Some searching has found various reports of the same thing happening on older WebKit versions, which makes sense, as that's what Phantom is based on. 

The [current implementation of `isExtraneousPopstateEvent`](https://github.com/ReactTraining/history/blob/8d98aec1366b86e3832dbfd7dcc480d1f9a5275c/modules/DOMUtils.js#L56-L58) does not appear to have anything *specific* to PhantomJS:

```js
export const isExtraneousPopstateEvent = event =>
  event.state === undefined &&
  navigator.userAgent.indexOf('CriOS') === -1
```

I added some prints to this code and can see that `event.state` is `null`, not `undefined`.

In decidim/decidim#1458, they chose to use the `isTrusted` property, so that might be a direction? I see that the event only has these properties (`state,returnValue,timeStamp,eventPhase,target,defaultPrevented,srcElement,type,clipboardData,cancelable,currentTarget,bubbles,cancelBubble`), so maybe even just checking for the presence of that property would be enough.

For reference, my Phantom's user agent is `Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/538.1 (KHTML, like Gecko) PhantomJS/2.1.1 Safari/538.1`.

**Comment by shepmaster** [comment]
As a local workaround, I've wrapped the history object and am passing it to redux-little-router:

```js
import { createBrowserHistory } from 'history';

// On page load, PhantomJS triggers two POP actions, while the newer
// spec says to only generate one. This causes double loads in our
// tests. This is inefficient and causes tests to fail when the DOM is
// loaded once, we start to interact with it, then the second load
// finishes and reverts our changes.
//
// Until this is fixed upstream, we ignore the first POP action that
// occurs if it looks like we are inside PhantomJS.
export default function createBrowserHistorySafeForTesting(args) {
  const history = createBrowserHistory(args);
  const originalListen = history.listen;

  let popCallsToBurn = howManyPopCallsToBurn();

  history.listen = realCallback => {
    const callback = (location, action) => {
      if (action === 'POP') {
        if (popCallsToBurn === 0) {
          realCallback(location, action);
        } else {
          popCallsToBurn--;
        }
      } else {
        realCallback(location, action);
      }
    };

    originalListen(callback);
  };

  return history;
}

function howManyPopCallsToBurn() {
  const looksLikePhantom =
        window &&
        window.navigator &&
        window.navigator.userAgent &&
        window.navigator.userAgent.indexOf('PhantomJS/2.1') !== -1;

  return looksLikePhantom ? 1 : 0;
}
```

**Comment by joshjg** [comment]
Just ran into this as well.

**Comment by mjackson** [comment]
Closing this issue due to inactivity. If it's still important to you, please comment.

**Comment by shepmaster** [comment]
@mjackson considering there's an open PR that purports to fix this issue (#520), it seems that other people are indeed experiencing the same problems...
